src/librustc_typeck/check/pat.rs

```diff
@@ -614,8 +618,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
 
         // Type-check the path.
-        let pat_ty = self.instantiate_value_path(segments, opt_ty, res, pat.span, pat.hir_id).0;
-        self.demand_suptype(pat.span, expected, pat_ty);
+        let (pat_ty, pat_res) =
+            self.instantiate_value_path(segments, opt_ty, res, pat.span, pat.hir_id);
+
+        let path = hir::print::to_string(tcx.hir(), |s| s.print_qpath(qpath, false));
+        if let Some(mut e) = self.demand_suptype_diag(pat.span, expected, pat_ty) {
+            if let Some(span) =
+                pat_res.opt_def_id().and_then(|def_id| self.tcx.hir().span_if_local(def_id))
+            {
+                e.span_label(span, &format!("`{}` defined here", path));
+                if segments.len() == 1 {
+                    e.note(&format!(
+                        "`{}` is interpreted as a {}, not a new binding",
```

Comment:
Would be good to make the logic uniform with `check_match::const_not_var` (and below) to the extent possible.